File: DataStructures/python/Queue/Queue.py
class Queue:

    def __init__(self, capacity) -> None:
        self.queue = [None for _ in range(capacity)]
        self.size = 0
        self.capasity = capacity
        self.front = 0
        self.rear = capacity - 1  
    
    def is_full(self):
        return self.size == self.capasity

    def is_empty(self):
        return self.size == 0

    def enqueue(self, item):
        if self.is_full():
            return "over flow"

        self.rear = (self.rear + 1) % self.capasity
        self.queue[self.rear] = item
        self.size += 1

    def dequeue(self):
        if self.is_empty():
            return "empty"
        
        output = self.queue[self.front]
        self.queue[self.front] = None
        self.front = (self.front + 1) % self.capasity
        self.size -= 1
        return output

    def get_front(self):
        return self.queue[self.front]

    def get_rear(self):
        return self.queue[self.rear]
    
    def get_size(self):
        return self.size


queue = Queue(3)
assert queue.get_size() == 0

queue.enqueue(5)
print(queue.queue)

assert queue.get_size() == 1
assert queue.get_front() == 5
assert queue.get_rear() == 5

queue.enqueue(6)
print(queue.queue)

assert queue.get_size() == 2
assert queue.get_front() == 5
assert queue.get_rear() == 6

queue.enqueue(7)
print(queue.queue)

assert queue.get_size() == 3
assert queue.get_front() == 5
assert queue.get_rear() == 7
assert queue.is_full() == True
assert queue.enqueue(8) == "over flow"

assert queue.dequeue() == 5
print(queue.queue)
assert queue.get_size() == 2

queue.enqueue(15)
print(queue.queue)
assert queue.get_size() == 3

assert queue.dequeue() == 6
print(queue.queue)
assert queue.dequeue() == 7
print(queue.queue)
assert queue.dequeue() == 15
print(queue.queue)
assert queue.is_empty() == True



    

File: DataStructures/php/Queue/Queue.php
<?php

class Queeu
{

    private int $capacity;
    private int $size;
    private int $front;
    private int $rear;
    private array $queue;

    public function __construct(int $capacity)
    {
        $this->capacity = $capacity;
        $this->size = 0;
        $this->front = 0;
        $this->rear = $capacity - 1;
        $this->setQueue();
    }

    private function setQueue()
    {
        for ($i = 0; $i < $this->capacity; $i++) {
            $this->queue[] = null;
        }
    }

    public function isEmpty()
    {
        return $this->size === 0;
    }

    public function isFull()
    {
        return $this->size === $this->capacity;
    }

    public function enQueue($item)
    {
        if($this->isFull())
        {
            throw new OverflowException('queue is full');
        }

        $this->rear = ($this->rear += 1) % $this->capacity;

        $this->queue[$this->rear] = $item;
        $this->size += 1;
    }

    public function deQueue()
    {
        if($this->isEmpty())
        {
            throw new Exception('queue is empty');
        }

        $item = $this->queue[$this->front];
        $this->front = ($this->front += 1) % $this->capacity;
        $this->size -= 1;
        return $item;

    }

    public function getFront()
    {
        return $this->queue[$this->front];
    }

    public function getRear()
    {
        return $this->queue[$this->rear];
    }

    public function getSize()
    {
        return $this->size;
    }

    public function __toString()
    {
        $output = '';
        foreach ($this->queue as $item) {
            $output .= $item ? "$item" :  "null";
            $output .= " ";
        }
        return $output . "\n";
    }
}

$queue = new Queeu(3);


echo $queue;
$queue->enQueue(1);


$queue->enQueue(2);
echo $queue;


$queue->enQueue(3);
echo $queue;

try{
    $queue->enQueue(1);

}catch(Exception $e){
    echo $e->getMessage();
    echo PHP_EOL;
}


echo $queue->deQueue() . "\n";
$queue->enQueue(10);
echo $queue;


echo $queue->deQueue() . "\n";
echo $queue;       

try{
    echo $queue->deQueue() . "\n";
    echo $queue->deQueue() . "\n";
    echo $queue->deQueue() . "\n";
    
}catch(Exception $e)
{
    echo $e->getMessage();
    echo PHP_EOL; 
}

$queue->enQueue(100);
$queue->enQueue(200);
echo $queue;
echo 'front = ' . $queue->getFront() . "\n";
echo 'rear = ' . $queue->getRear() . "\n";
echo 'size = ' . $queue->getSize() . "\n";






